src/workerd/api/actor-state.c++

```diff
@@ -640,6 +640,12 @@ jsg::Value DurableObjectStorage::transactionSync(jsg::Lock& js, jsg::Function<js
     sqlite->run(SqliteDatabase::TRUSTED, kj::str("SAVEPOINT _cf_sync_savepoint_", depth));
     return js.tryCatch([&]() {
       auto result = callback(js);
+      if (!sqlite->run("SELECT * FROM pragma_foreign_keys(), pragma_foreign_key_check() WHERE foreign_keys = TRUE;").isDone()) {
```

Comment:
If we are really going to run this query on every transaction we should make it a prepared statement.

But can we avoid it entirely, at least in the common case where it won't apply? What if we prohibit deferred foreign key checks (in the authorizer)? Or can we use the authorizer to detect whether someone has enabled deferred checks, and only then start doing this check here?

Reply:
Yeah that's the primary reason I pushed this as a draft PR. I want this:

> can we use the authorizer to detect whether someone has enabled deferred checks, and only then start doing this check here?

But I don't know how to pass context between those callsites? If we could do _that_, then we could also provide better error messages (instead of `not authorized`, we could give `not authorized: SQLITE_CREATE_TRIGGER disallowed for name '_cf_KV'`). Something I've wanted to add but didn't know how...

Reply:
So my other question is, where should the prepared statements be stored? Right now `ActorSqlite` has a couple (`beginTxn`,`commitTxn`), and I see `SqlStorage` having `pragmaPageCount` and `pragmaGetMaxPageCount`, and using the (currently private) `execMemoized` to store them on first use, but this is the generic `ActorState` class.

Gut feel says I want to replace `KJ_IF_MAYBE(sqlite, cache->getSqliteDatabase())` on line 633 with some kind of dynamic cast of `ActorCacheInterface` down to `ActorSqlite` so I can move `transactionSync` on to it, and store prepared statements there. Or, add `transactionSync` to `ActorCacheInterface`, and throw an error if you try to do it on a non-SQL-backed Cache... I might try that.

Reply:
Ok so I tried and failed. Something to do with JSG and template specialisation which I don't understand (error in commit description here: https://github.com/cloudflare/workerd/commit/46936690ecaf32a06c43948ec861add8ed9cd601)